src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -169,20 +173,57 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                 {
                     if (NetEventSource.IsEnabled) Trace($"Failed to send request body: {e}");
 
-                    // Cancel the stream before we set _requestCompletionState below.
-                    // Otherwise, a response stream reader may race with the actual Cancel.
-                    Cancel();
+                    bool signalWaiter = false;
 
                     lock (SyncObject)
                     {
                         Debug.Assert(_requestCompletionState == StreamCompletionState.InProgress, $"Request already completed with state={_requestCompletionState}");
 
-                        _requestCompletionState = StreamCompletionState.Failed;
+                        if (_requestBodyAbandoned)
+                        {
+                            // See comments on _requestBodyAbandoned.
+                            // In this case, the request is still considered successful and we do not want to send a RST_STREAM, 
+                            // and we also don't want to propagate any error to the caller, in particular for non-duplex scenarios.
+                            Debug.Assert(_responseCompletionState == StreamCompletionState.Completed);
+                            _requestCompletionState = StreamCompletionState.Completed;
+                            Complete();
+                            return;
+                        }
+                        else
+                        {
+                            _requestCompletionState = StreamCompletionState.Failed;
 
-                        // Cancel above should ensure that the response is either Completed or Failed now.
-                        Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                            // This duplicates some of the existing logic in Cancel.
+                            // Cancel takes the lock, so we can't call it here while holding the lock.
+                            // We can't call Cancel before taking the lock, because we don't want to cancel in the _requestBodyAbandoned case.
+                            // But we also don't want to call Cancel after releasing the lock, because that can cause a race
+                            // where the client sees the request body has failed but the response body hasn't.
+                            // We should consider how to refactor this logic to avoid code duplication.
```

Comment:
> We should consider how to refactor this logic to avoid code duplication.

Yes, please.  Can you open an issue for that?  All this duplication makes me nervous.

Reply:
That said, this looks like a direct copy of a good chunk of the code that's done in Cancel.  Can't we just now create a function that both Cancel and this call?

Reply:
Sure, I'll do that.